corehq/sql_db/tests/test_jsonops.py

```diff
@@ -0,0 +1,139 @@
+
+
+def test_JsonSet_sql():
+    expr = ops.JsonSet(data_ref, ["items", 0, "abc"], [1, 2, 3])
+    eq(as_sql(expr), "jsonb_set(data, ('{items,0,abc}'), [1, 2, 3], true)")
+
+
+def test_JsonSet_do_not_create_missing():
+    expr = ops.JsonSet(data_ref, ["items"], [1, None], create_missing=False)
+    eq(as_sql(expr), "jsonb_set(data, ('{items}'), [1, null], false)")
+
+
+def test_JsonSet_illegal_field():
+    with assert_raises(ValueError):
+        ops.JsonSet(data_ref, ["illegal'field"], 1)
+
+
+def test_JsonSet_with_expression():
+    select = RawSQL("SELECT '{}'::jsonb", [])
+    expr = ops.JsonSet(data_ref, ["field"], select)
+    eq(as_sql(expr), "jsonb_set(data, ('{field}'), (SELECT '{}'::jsonb), true)")
+
+
+def test_nested_operations():
+    expr = ops.JsonSet(
+        ops.JsonDelete(data_ref, "things"),
+        ["items"],
+        ops.JsonGet(data_ref, "things"),
+    )
+    eq(as_sql(expr), "jsonb_set(data::jsonb - '{things}'::text[], ('{items}'), data->'things', true)")
+
+
+class TestJsonOpsEvaluation(TestCase):
+
+    def test_JsonDelete(self):
+        value = eval_json_op(ops.JsonDelete(data_ref, "score"))
+        self.assertEqual(value, {"things": [1, 2, 3], "other": "value"})
+
+    def test_JsonSet(self):
+        value = eval_json_op(ops.JsonSet(data_ref, ["things", 1], {"b": 2}))
+        self.assertEqual(value, {
+            "things": [1, {"b": 2}, 3],
+            "other": "value",
+            "score": 3,
+        })
+
+    def test_JsonGet(self):
+        value = eval_json_op(ops.JsonGet(data_ref, "things"))
+        self.assertEqual(value, [1, 2, 3])
+
+    def test_nested_operations(self):
+        value = eval_json_op(ops.JsonSet(
+            ops.JsonDelete(data_ref, "things"),
+            ["items"],
+            ops.JsonGet(data_ref, "things"),
+        ))
+        self.assertEqual(value, {
+            "items": [1, 2, 3],
+            "other": "value",
+            "score": 3,
+        })
+
+
+class _UnquotedRef(Ref):
+
+    def as_sql(self, compiler, connection):
+        return self.refs, []
+
+
+data_ref = _UnquotedRef("data", None)
+
+
+def as_sql(expression):
+    compiler = SQLCompiler(None, connection, DEFAULT_DB_ALIAS)
+    sql, params = compiler.compile(expression)
+    return sql % tuple(params)
+
+
+def eval_json_op(expression):
+    # raw CTE is a hack to evaluate a query without an existing model
+    data = {"things": [1, 2, 3], "other": "value", "score": 3}
+    cte_sql = "SELECT %s::jsonb AS data"
+    data_cte = raw_cte_sql(cte_sql, [json.dumps(data)], {"data": JSONField()})
+    data_cte.query.model = _JsonModel
+    data_cte.query.default_cols = []
+    data_cte.query.annotations = {}
+    data_cte.query.values_select = []
+    data_cte.query.annotation_select_mask = None
+    cte = With(data_cte)
+    qs = (
+        cte.queryset().with_cte(cte)
+        .annotate(val=expression)
+        .values_list("val", flat=True)
+    )
+    print(qs.query)
```

Comment:
Is this a debug print you meant to remove or useful test output?

Reply:
Useful test output.